plugin/dnstap/handler.go

```diff
@@ -20,12 +20,20 @@ type Dnstap struct {
 	IncludeRawMessage bool
 	Identity          []byte
 	Version           []byte
+	Extra             []byte
 }
 
 // TapMessage sends the message m to the dnstap interface.
 func (h Dnstap) TapMessage(m *tap.Message) {
 	t := tap.Dnstap_MESSAGE
-	h.io.Dnstap(&tap.Dnstap{Type: &t, Message: m, Identity: h.Identity, Version: h.Version})
+	dt := &tap.Dnstap{
+		Type: &t,
+		Message: m,
+		Identity: h.Identity,
+		Version: h.Version,
+		Extra: h.Extra,
```

Comment:
what you implemented here is assigning a preconfigured text to `Extra` (which can be useful as well), whereas https://github.com/coredns/coredns/issues/6092 suggests extracting and assigning value of preconfigured metadata label.

The metadata value is normally created by other plugin, e.g. if you specify label "forward/upstream", it is expected that plugin 'forward' should provide this value. The plugin which provides metadata can generate different values for different queries, and this is the key difference from what you have implemented.

Please get familiar with plugin metadata https://github.com/coredns/coredns/tree/master/plugin/metadata as well as the package https://github.com/coredns/coredns/tree/master/plugin/pkg/replacer which can make your work easier

I also recommend to look at plugins 'rewrite' and 'log' as examples of plugins which read metadata from other plugins.

Reply:
Thanks for your kind review! I have introduced the `replacer` package in my new commits.

Since the replacer requires context, status (`request.Request`) and rr (`*dnstest.Recorder`), I have added a new function called `TapMessageWithMetadata` to receive these context objects and process the metadata replacement accordingly. 

The old function `TapMessage` has not been removed, as external plugins may have dependencies on it.